pkg/webhook/cluster/mutating_webhook.go

```diff
@@ -58,18 +67,122 @@ func (a *ManagedClusterMutatingAdmissionHook) Admit(req *admissionv1beta1.Admiss
+		return status
+	}
+
+	patch, err := json.Marshal(jsonPatches)
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusInternalServerError, Reason: metav1.StatusReasonInternalError,
+			Message: err.Error(),
+		}
+		return status
 	}
 
+	status.Patch = patch
+	pt := admissionv1beta1.PatchTypeJSONPatch
+	status.PatchType = &pt
 	return status
 }
 
+// processTaints generates json patched for cluster taints
+func (a *ManagedClusterMutatingAdmissionHook) processTaints(managedCluster *clusterv1.ManagedCluster, oldManagedClusterRaw []byte) ([]jsonPatchOperation, *admissionv1beta1.AdmissionResponse) {
+	status := &admissionv1beta1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	if len(managedCluster.Spec.Taints) == 0 {
+		return nil, status
+	}
+
+	var oldManagedCluster *clusterv1.ManagedCluster
+	if len(oldManagedClusterRaw) > 0 {
+		cluster := &clusterv1.ManagedCluster{}
+		if err := json.Unmarshal(oldManagedClusterRaw, cluster); err != nil {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusInternalServerError, Reason: metav1.StatusReasonInternalError,
+				Message: err.Error(),
+			}
+			return nil, status
+		}
+		oldManagedCluster = cluster
+	}
+
+	var invalidTaints []string
+	var jsonPatches []jsonPatchOperation
+	now := metav1.NewTime(nowFunc())
+	for index, taint := range managedCluster.Spec.Taints {
+		originalTaint := findTaintWithKey(oldManagedCluster, taint.Key)
+		switch {
+		case originalTaint == nil:
+			// new taint
+			if !taint.TimeAdded.IsZero() {
+				invalidTaints = append(invalidTaints, taint.Key)
+				continue
+			}
+			jsonPatches = append(jsonPatches, newTaintTimeAddedJsonPatch(index, now.Time))
+		case originalTaint.Value == taint.Value && originalTaint.Effect == taint.Effect:
+			// no change
+			if !originalTaint.TimeAdded.Equal(&taint.TimeAdded) {
+				invalidTaints = append(invalidTaints, taint.Key)
+			}
+		default:
+			// taint's value/effect has changed
+			if !taint.TimeAdded.IsZero() {
+				invalidTaints = append(invalidTaints, taint.Key)
+				continue
+			}
+			jsonPatches = append(jsonPatches, newTaintTimeAddedJsonPatch(index, now.Time))
+		}
+	}
+
+	if len(invalidTaints) == 0 {
+		return jsonPatches, status
+	}
+
+	status.Allowed = false
+	status.Result = &metav1.Status{
+		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+		Message: fmt.Sprintf("It is not allowed to set TimeAdded of Taint %q.", strings.Join(invalidTaints, ",")),
+	}
+	return nil, status
+}
+
 // Initialize is called by generic-admission-server on startup to setup initialization that managedclusters webhook needs.
 func (a *ManagedClusterMutatingAdmissionHook) Initialize(kubeClientConfig *rest.Config, stopCh <-chan struct{}) error {
 	// do nothing
 	return nil
 }
+
+func findTaintWithKey(managedCluster *clusterv1.ManagedCluster, key string) *clusterv1.Taint {
```

Comment:
put this as a common helper func. I feel you will also need it in controller's code

Reply:
and the func is better to be `FindTaintByKey(taints, key)`

Reply:
Done.

Reply:
Did you push the commit? I do not see the change here yet.

Reply:
The change is in place now.